File: lab/javascript/arrays_e_listas/soma_array.js
class ArrayCustom {
  constructor(tamanho) {
    this.tamanho = tamanho;
    this.array = [];
  }

  toString() {
    return `ArrayCustom(${JSON.stringify(this.array)})`;
  }

  add(n) {
    if (this.array.length >= this.tamanho) {
      throw new Error("Limite máximo atingido para o array");
    }
    this.array.push(n);
  }

  delete() {
    if (this.array.length <= 0) {
      throw new Error("Lista vazia para remover");
    }
    this.array.pop();
  }

  sumArray() {
    return this.array.reduce((a, b) => a + b, 0);
  }
}

// -------------------- Teste --------------------
const lista = new ArrayCustom(5);

for (let i = 1; i <= 5; i++) {
  lista.add(i);
}

console.log("Array:", lista.toString());
console.log("Soma dos elementos:", lista.sumArray());

try {
  lista.add(6);
} catch (e) {
  console.log("Erro esperado:", e.message);
}

lista.delete();
console.log("Após remover um elemento:", lista.toString());
console.log("Soma atual:", lista.sumArray());

File: lab/c/arrays_e_listas/soma_array.c
#include <stdio.h>
#include <stdlib.h>

typedef struct {
    int tamanho;
    int *array;
    int count;
} Array;

void init(Array *a, int tamanho) {
    a->tamanho = tamanho;
    a->array = malloc(sizeof(int) * tamanho);
    a->count = 0;
}

void add(Array *a, int n) {
    if (a->count >= a->tamanho) {
        printf("Erro: Limite máximo atingido para o array\n");
        return;
    }
    a->array[a->count++] = n;
}

void delete(Array *a) {
    if (a->count <= 0) {
        printf("Erro: Lista vazia para remover\n");
        return;
    }
    a->count--;
}

int sum_array(Array *a) {
    int soma = 0;
    for (int i = 0; i < a->count; i++) {
        soma += a->array[i];
    }
    return soma;
}

void print_array(Array *a) {
    printf("ArrayCustom([");
    for (int i = 0; i < a->count; i++) {
        printf("%d", a->array[i]);
        if (i < a->count - 1) printf(", ");
    }
    printf("])\n");
}

int main(void) {
    Array lista;
    init(&lista, 5);

    // Adiciona de 1 a 5
    for (int i = 1; i <= 5; i++) {
        add(&lista, i);
    }

    printf("Array: ");
    print_array(&lista);
    printf("Soma dos elementos: %d\n", sum_array(&lista));

    // Testando erro (adicionar além do limite)
    add(&lista, 6);

    // Removendo e somando de novo
    delete(&lista);
    printf("Após remover um elemento: ");
    print_array(&lista);
    printf("Soma atual: %d\n", sum_array(&lista));

    free(lista.array);
    return 0;
}
